### PR TITLE
Always test explicit names with name predicate when ignore unavailable 

### PR DESCRIPTION
**Pull request by albertzaharovits**
### Background

The Security filter rewrites requests to remove some **non-wildcard** names
when the `ignore_unavailable` request option is `true`. See: https://github.com/elastic/elasticsearch/blob/3fd338ffd15d7d7cafc15b6405d11e64a4641225/server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstractionResolver.java#L103 and https://github.com/elastic/elasticsearch/blob/995f578d83970d85df9fd41838a8b5c4e69dfb77/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java#L785

It **must** remove unauthorized names.
But it doesn't have to remove **authorized** names that:
 * are datastreams, in case the request doesn't handle datastreams (Core code ignores such datastreams already)
 * are backing indices of datastreams in case the request doesn't handle datastreams and the user only has permissions on the datastream, rather than on the index itself (it is probably a bug that such indices are currently removed)
 * are missing (Core code ignores missing names already, and removing them on the Security filter is causing the bug described here https://github.com/elastic/elasticsearch/issues/90215)

### Proposal

Only remove **unauthorized** explicit names (existing or not) when `ignore_unavailable` is `true`,
in the Security filter, and leave the other explicit names, which are authorized, to be processed on
the Core side.

The first step towards that, which is what this PR does, is to always apply the "is-authorized" predicate
on the explicit names, rather than, in some cases, verify that the explicit name is among the
**existing** and authorized indices subset.

### CODE REVIEW

**Comment by jakelandis** [comment]
I'm having a hard time connecting the dots between this PR and the proposal. On the surface it seems that this PR is just just removing an optimization (i.e. if we have fully resolved the supplier then just use the inner set as a cache to avoid work when checking for contains an item/collection ), so I am having a hard time seeing how the lack of optimization leads into a desired functional change. 

Also it is difficult to trace back to where all we call the contains/containsAll method to determine the impact of the removal of the optimization.

**Comment by ywangd** [comment]
@jakelandis 

> where all we call the contains/containsAll method to determine the impact of the removal of the optimization.

The `contains` method is called at
https://github.com/elastic/elasticsearch/blob/e68e28e4cb983cf06be5612b38078bc7d9e02f48/server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstractionResolver.java#L103

The `containsAll` method is never used. It is included mainly for the `Set` interface.

For pure performance reason, I don't think we should remove the optimisation. The predicate is most likley more expensive since it internally has a `Map.get` call and a few other things. So it's going to be slower.

But I think @albertzaharovits needs it for the proposal here because the internal `authorizedIndices` set only contains authorized names that **exist** in the cluster state. It does *not* contain names that are authorized but absent from the cluster state (the absence could be temprorary because the index is being recovered). So the difference is that `authorizedIndices.contains` will return `false`, but `predicate.test` will return `true` for such index name. This subtle difference is what causing the issue in #90215.

Though this change will fix the specific issue of #90215, I feel it is a bit too obscure. It breaks the contract of the `Set` interface, i.e. `contains` returns `true`, but neither `iterator` nor `size` includes such element. That is weird at least. It also does not help when the request index name is a wildcard (e.g. `.kibana*`). The perfomance implication may not be much, but I cannot be sure either. Is it blocking your other work? If not, we may choose to defer it and look at it again once larger scale restructure is done?

**Comment by albertzaharovits** [comment]
> > I'm having a hard time connecting the dots between this PR and the proposal. On the surface it seems that this PR is just just removing an optimization (i.e. if we have fully resolved the supplier then just use the inner set as a cache to avoid work when checking for contains an item/collection ), so I am having a hard time seeing how the lack of optimization leads into a desired functional change.

> I think @albertzaharovits needs it for the proposal here because the internal authorizedIndices set only contains authorized names that exist in the cluster state. It does not contain names that are authorized but absent from the cluster state (the absence could be temprorary because the index is being recovered). So the difference is that authorizedIndices.contains will return false, but predicate.test will return true for such index name. This subtle difference is what causing the issue in https://github.com/elastic/elasticsearch/issues/90215.

What Yang says. The proposal is to "Only remove unauthorized explicit names", but the optimization that this PR takes away makes it that unauthorized **or missing** explicit names are removed. The optimisation is incompatible with the bug fix proposal.

> For pure performance reason, I don't think we should remove the optimisation. The predicate is most likley more expensive since it internally has a Map.get call and a few other things. So it's going to be slower.

I think it's worth pointing out that the predicate is only evaluated for explicit names. In general I think a performance penalty on the order of the explicit names in the request is not something we should worry about too much. Moreover, this inneficiency pales in comparison, to the check in https://github.com/elastic/elasticsearch/blob/e68e28e4cb983cf06be5612b38078bc7d9e02f48/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java#L510 which runs AGAIN on the explicit names, and the results from wildcard expansions.

> Though this change will fix the specific issue of https://github.com/elastic/elasticsearch/issues/90215, I feel it is a bit too obscure. It breaks the contract of the Set interface, i.e. contains returns true, but neither iterator nor size includes such element.

I agree. I think the `AuthorizedIndices` class should not implement the `Set` interface anymore. I can implement this if you consent.

> It also does not help when the request index name is a wildcard (e.g. .kibana*).

True, but I think this has been acknowledged https://github.com/elastic/elasticsearch/issues/90215#issuecomment-1273249327 (last line: "Make the Security filter NOT remove authorized names for missing indices when ignore_unavailable=true; this will fix this bug issue specifically, but if instead the request used wildcards rather than explicit names, it would still be possible to erroneously report that no such indices exist when in fact the cluster state is not yet recovered.")

> Is it blocking your other work? If not, we may choose to defer it and look at it again once larger scale restructure is done?

It blocks fixing a bug that I'm assigned to, which I don't know how to fix otherwise, and which I think makes ES appear inconsistent (so it's pretty serious), though the circumstances are particular (require usage of service account tokens).
It doesn't block the indices expression resolver refactoring.

**Comment by ywangd** [comment]
> I think it's worth pointing out that the predicate is only evaluated for explicit names. In general I think a performance penalty on the order of the explicit names in the request is not something we should worry about too much.

Fair enough. I had concrete names as a result of wildcard expansion in mind. But now I think this code should only apply to concrete index names directly supplied by end users. In that case, it is reasonable to expect the number of concrete names is small enough to not be concerned. I think we can also have something like

```java
if (authorizedIndices != null && authorizedIndices.contains(name)) {
  return true;
} else {
  return predicate.test(name);
}
```
which will run two tests if `contains` returns `false`. But `contains` returning false is an edge case already, so running an extra `predicate.test` is not really an issue. It helps largely maintain the existing performance profile.

> I think the AuthorizedIndices class should not implement the Set interface anymore. I can implement this if you consent.

That'd be great. This was my biggest gripe. It also helps get rid of the unused method like `containsAll`.

> True, but I think this has been acknowledged https://github.com/elastic/elasticsearch/issues/90215#issuecomment-1273249327

I forgot about that. Thanks for reminding me. 

Btw, since we are aming to have a single implementation between core and security, have you considered how to unify them (different behaviours for `ignore_unavailable` on explicit names) when the time comes? I guess it can be parameterized? Or something else?

In summary, I am happy for you to proceed as discussed above. Thanks!

**Comment by albertzaharovits** [comment]
> Btw, since we are aming to have a single implementation between core and security, have you considered how to unify them (different behaviours for ignore_unavailable on explicit names) when the time comes? I guess it can be parameterized? Or something else?

I hope in the end the method from Core will be parameterized, and called by the Security filter. The handling of explicit names is completely different in Security, and to make it easier to observe, I endeavor to pull it out of the wildcard resolver.

**Comment by albertzaharovits** [comment]
> > I think the AuthorizedIndices class should not implement the Set interface anymore. I can implement this if you consent.

> That'd be great. This was my biggest gripe. It also helps get rid of the unused method like containsAll.

I have updated the PR to that effect. It's now a big boy.

@ywangd I like the suggestion from:
> if (authorizedIndices != null && authorizedIndices.contains(name)) {
  return true;
} else {
  return predicate.test(name);
}

but I've not incorporated it for now. I feel like it's a bit too smart to squeeze it in this PR. But I'm open to reconsider if you disagree.

---------------

So I think this PR should be merged if (cc @jakelandis):
 * we want to fix https://github.com/elastic/elasticsearch/issues/90215 in the current architecture of a Security filter that rewrites requests scoped by the user's access level
 * we don't have a better alternative than the proposed solution that entails we continue to expand wildcards to existing indices, but explicit name authorization is indiferent to indices existing or not

**Review by jakelandis — COMMENTED** [review]
The changes here look good. I like the use of LazyInitializable to wrap the set instead of implementing the set interface for reasons previous discussed and the unique type makes it is easier to track the usages of when/where it used and fully resolved or not. 

However, maybe I am just confused (this all very complex), but I am still not fully understanding how this change fixes the initialization issue. 

> because the internal authorizedIndices set only contains authorized names that exist in the cluster state. It does not contain names that are authorized but absent from the cluster state (the absence could be temprorary because the index is being recovered). So the difference is that authorizedIndices.contains will return false, but predicate.test will return true for such index name. 

I don't see how the fully resolved set (i.e. allAuthorizedAndAvailable.getOrCompute()) would ever differ from singular predicate check (isAuthorized.test)). In both cases the index must be present in the view of cluster state to prevent it from being removed from the view that core will see. Both cases (full set / singular predicate) use the same view of the cluster state. When looking at the fully resolved set, the index will only be included in the results iff it is found in cluster state. When looking at the singular predicate, it will be only return true if the it is found in the cluster state (we explicitly look up the name and [return false](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java#L788) if not found.)  So I guess I don't understand how this change influences what will be presented to core vs. not.

**Comment by ywangd** [comment]
> I don't see how the fully resolved set (i.e. allAuthorizedAndAvailable.getOrCompute()) would ever differ from singular predicate check (isAuthorized.test)). In both cases the index must be present in the view of cluster state to prevent it from being removed from the view that core will see. Both cases (full set / singular predicate) use the same view of the cluster state. When looking at the fully resolved set, the index will only be included in the results iff it is found in cluster state. When looking at the singular predicate, it will be only return true if the it is found in the cluster state (we explicitly look up the name and [return false](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java#L788) if not found.) So I guess I don't understand how this change influences what will be presented to core vs. not.

I think @jakelandis is absolutely right on this point. The predicate internally relies on the same cluster state view. As such, it won't give a different answer. I glossed over this important detail 🤦 It also shows that we need new tests for the change.

This might be a deal breaker ... Trying to work around it would lead to many cascading changes. It also has to ignore data stream (this reminds me that our index authorization is no longer purely string based). The benefit feels small compare to the complexities and compromises[^1]. It makes me retreat to my previous position of deferring it. Maybe I can be convinced again if there are creative ways to make it work. For now, this looks hard to converge because the fundamental underlying issue is security filter changes the order of execution for index resolving. What we are trying to have is an edge case point solution while the larger problem remains. The "alias pointing to closed index" problem can also be attributed to this overall issue. I wonder whether our long term goal is still have security filter plug into core (sorry I lost track on this)? If so, that might be the actual solution to this. But it is very hard change by itself, so I am not sure whether we can rely on that either. Sorry for the back and forth. This is more conflicting than I expected. I am willing to hear out other opinions. Thanks!

[^1]: It does not work with wildcards and probably not with explicit names matching backing indices either. The current behaviour, though not ideal, is more self-consistent, i.e. if the index is not available at the time we check cluster state, then it is not available. The workaround is also simple, just use `ignore_unavailable=false` which is the **default**. The proposed behaviour feels pretty hard to explain and document.

**Comment by albertzaharovits** [comment]
> > I don't see how the fully resolved set (i.e. allAuthorizedAndAvailable.getOrCompute()) would ever differ from singular predicate check (isAuthorized.test)). In both cases the index must be present in the view of cluster state to prevent it from being removed from the view that core will see. Both cases (full set / singular predicate) use the same view of the cluster state. When looking at the fully resolved set, the index will only be included in the results iff it is found in cluster state. When looking at the singular predicate, it will be only return true if the it is found in the cluster state (we explicitly look up the name and [return false](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java#L788) if not found.) So I guess I don't understand how this change influences what will be presented to core vs. not.

> I think @jakelandis is absolutely right on this point. The predicate internally relies on the same cluster state view. As such, it won't give a different answer. I glossed over this important detail 🤦 It also shows that we need new tests for the change.

This PR is not a fix for the bug! From the PR description:
> The first step towards that, which is what this PR does, is to always apply the "is-authorized" predicate
on the explicit names, rather than, in some cases, verify that the explicit name is among the
existing and authorized indices subset.

The predicate must be changed for the PR to be fixed. But this is going to be a follow-up PR. And that PR is going to be labeled `bug-fix` . And before that I was going to raise a discuss issue because there are two options for the new predicate that I would need your input to evaluate.
This PR is just a starting point. Look at the initial commit, from the version I initially asked review of: https://github.com/elastic/elasticsearch/pull/91618/commits/1a3e6479f31c6a80982c6cc0589301a77270f714 . How could have that change fixed the bug? The predicate test and the set test are equivalent today, otherwise the current code in `main` would be broken. I need to get this PR merged first, because I cannot change the predicate and keep the authorized set logic the same, because the is authorized check of "sometimes check set contains sometimes evaluate predicate" would be broken. 

Should we set a sync discussion to clarify the scope of this PR?

**Comment by albertzaharovits** [comment]
> This might be a deal breaker ... Trying to work around it would lead to many cascading changes. It also has to ignore data stream (this reminds me that our index authorization is no longer purely string based). The benefit feels small compare to the complexities and compromises[1](https://github.com/elastic/elasticsearch/pull/91618#user-content-fn-1-b689671c263724328c022f7a714f8624). It makes me retreat to my previous position of deferring it. Maybe I can be convinced again if there are creative ways to make it work. For now, this looks hard to converge because the fundamental underlying issue is security filter changes the order of execution for index resolving. What we are trying to have is an edge case point solution while the larger problem remains. The "alias pointing to closed index" problem can also be attributed to this overall issue. I wonder whether our long term goal is still have security filter plug into core (sorry I lost track on this)? If so, that might be the actual solution to this. But it is very hard change by itself, so I am not sure whether we can rely on that either. Sorry for the back and forth. This is more conflicting than I expected. I am willing to hear out other opinions. Thanks!

I'm sorry but I don't understand what is the newly discovered issue that is a deal-breaker here. I will propose we have a sync conv.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java:343]: I agree it's complicating reformatting but this is spotless.

**Comment by albertzaharovits** [comment]
Yang and I had a chat before in the day on this topic. Here are my take aways from the conversation:

* this PR is not a bug fix, it is only a refactoring
* we need to discuss on the "is authorized" predicate, used in the request expression rewriting, in order to fix the bug in #90215
* the new predicate will have to work on `String`s too (for missing indices), very similarly to how the `IndicesPermission#isActionGranted` test works
* I should change the method signature for `AuthorizationEngine#AuthorizedIndices#allAuthorizedAndAvailable` back to return a `Supplier<Set<String>>`, because the lazy trait is an implementation detail and the lack of thread-safetiness is not a concern

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:937]: I've ended up replacing the `LazyInitializable` with a `CachedSupplier`, because `CachedSupplier` implements the `Supplier` interface which I've used in the method signature (because the "lazy" trait is an implementation detail).

**Review by jakelandis — APPROVED** [review]
Thanks for the additional clarifications. The changes here look good. I like the use of more explicit data types and thinks this is a bit easier to follow now. A minor (non-blocking) naming suggestion, and would also want a +1 from @ywangd before merging.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/AuthorizationEngine.java
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/AuthorizationEngine.java:264]: > should this be isAuthorizedAndAvailable ?

Yeah, given the current implementation, I think that would be more accurate.

> For example, if the name were more generic then the same interface could be used to get all authorized but unavaible.

Yes, that's the goal with this interface.

I'll ammend the naming as suggested.

Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/AuthorizationEngine.java]: Pushed [57d5777](https://github.com/elastic/elasticsearch/pull/91618/commits/57d5777cf005d1618f12e9205094a87242a899b3) and [1e8dc98](https://github.com/elastic/elasticsearch/pull/91618/commits/1e8dc9881967b881550cd99b902d3ec63e5520d9)

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/AuthorizedIndicesTests.java
Comment (ywangd) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/AuthorizedIndicesTests.java:111]: Nit: Can we renamed the variable to `authorizedIndices`? The existing name `list` is rather confusing after this refactor. The same comment applies to two other methods (`testDataStreamsAreNotIncludedInAuthorizedIndices`, `testDataStreamsAreIncludedInAuthorizedIndices`) in this test class as well.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java
Comment (ywangd) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java:2242]: Based on the pattern in this test class, should we also have
```suggestion
            assertThat(authorizedIndices.all().get(), hasItem(i.getName()));
            assertThat(authorizedIndices.check(i.getName()), is(true));
```
?

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java
Comment (ywangd) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java:2255]: Similarly
```suggestion
            assertThat(authorizedIndices.all().get(), hasItem(i.getName()));
            assertThat(authorizedIndices.check(i.getName()), is(true));
```

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/RBACEngineTests.java
Comment (ywangd) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/RBACEngineTests.java:1490]: Nit: This assertion is not necessary anymore since the AuthorizedIndices is no longer a Set implementation and the type is guaranteed statically.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/RBACEngineTests.java
Comment (ywangd) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/RBACEngineTests.java:1535]: Same here. The assertion is not needed anymore.

**Review by ywangd — APPROVED** [review]
LGTM Thanks for working through this and providing clarification along the way. 

The `CustomAuthorizationEngine` needs to be updated once this PR is merged.

This is an idea for possible future refactoring (no need to do anything for this PR): I like the new `AuthorizedIndices` interface. It is rather unfortunate that it has to be decomposed into two parameters when it is used with `IndexAbstractionResolver`. The main reason is because `IndexAbstractionResolver` is in core. But does it have to be? As far as I can tell, it is only used by `ResolveIndexAction`. Everything else in core uses `IndexNameExpressionResolver`. Is there anything special about `ResolveIndexAction` to justify it being different? If not, we could replace its usage and move `IndexAbstractionResolver` to xpack. This way, we can not only use the `AuthorizedIndices` interface everywhere but also likely be able to improve the interface to enforce its internal consistency, i.e. all names from the supplier should pass `check`, but not the other way around. We could even choose to hide the `Supplier` usage so that `AuthorizedIndices#all` simply returns a `Set`.

**Comment by albertzaharovits** [comment]
>  I like the new AuthorizedIndices interface. It is rather unfortunate that it has to be decomposed into two parameters when it is used with IndexAbstractionResolver. The main reason is because IndexAbstractionResolver is in core. But does it have to be? As far as I can tell, it is only used by ResolveIndexAction. Everything else in core uses IndexNameExpressionResolver. Is there anything special about ResolveIndexAction to justify it being different? If not, we could replace its usage and move IndexAbstractionResolver to xpack.

There shouldn't be anything special about the `ResolveIndexAction` for it to use the Security's implementation (which confusingly is packaged in Core) of the resolver, i.e. the `IndexAbstractionResolver`; though I haven't looked closely.
In any case, the objective is to use a single implementation for the resolver, the one Core uses, i.e. `org.elasticsearch.cluster.metadata.IndexNameExpressionResolver`. I'll hold off refactoring the `IndexAbstractionResolver` because it's very likely to be removed soon.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/AuthorizedIndicesTests.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/AuthorizedIndicesTests.java]: Pushed [d8953b2](https://github.com/elastic/elasticsearch/pull/91618/commits/d8953b2449110a85d65af30d231ed556c57e9c00)

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java:2242]: Pushed [8823f80](https://github.com/elastic/elasticsearch/pull/91618/commits/8823f80e83494043d2a573be3f50e4e18eabb582)

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java:2256]: Pushed [8823f80](https://github.com/elastic/elasticsearch/pull/91618/commits/8823f80e83494043d2a573be3f50e4e18eabb582)

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/RBACEngineTests.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/RBACEngineTests.java]: Pushed [531245e](https://github.com/elastic/elasticsearch/pull/91618/commits/531245e18824215ec0f6f316a0bfc508af2bd895)

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/RBACEngineTests.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/RBACEngineTests.java]: Pushed [531245e](https://github.com/elastic/elasticsearch/pull/91618/commits/531245e18824215ec0f6f316a0bfc508af2bd895)

**Comment by albertzaharovits** [comment]
> The CustomAuthorizationEngine needs to be updated once this PR is merged.

Pushed [209f101](https://github.com/elastic/elasticsearch/pull/91618/commits/209f10118d4ae79503d475abb6d7790a03242206).

For posterity, the example plugins, of which `security-authorization-engine` is one instance, are not tested in PR CI tests.
They are tested in a separate periodic CI job:  https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request+example-plugins/ .

If anything about the `AuthorizationEngine` interface changes, almost always the plugin's implementation, i.e. `CustomAuthorizationEngine`, also needs to be adjusted. You can use the following command to compile the `security-authorization-engine` plugin `(cd plugins/examples && ./gradlew :security-authorization-engine:compileJava --include-build ../..)` ; I also had to set `org.gradle.jvmargs=-Xmx4g` to my `~/.gradle/gradle.properties` file in order to avoid `OutOfMemoryError`s during the build.

**Pull request merged**
